File: projects/plateau-sniper-vs-guardman/unity/PlateauSniperGame/Assets/Scripts/Network/SignalingClient.cs
// =============================================================
//  SignalingClient.cs  (Unity 6)
//  PartyKit WebSocket シグナリングクライアント
//  ICE / SDP の交換のみ担当。位置データは一切通さない。
//  Websocket.Client (NuGet) を使用
// =============================================================

using System;
using System.Collections.Concurrent;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnityEngine;
using Websocket.Client;

namespace PlateauSniper.Network
{
    // ─── シグナリングメッセージ型 ─────────────────────────────

    public class SigPeerInfo
    {
        [JsonProperty("clientId")]    public string ClientId    { get; set; } = "";
        [JsonProperty("role")]        public string Role        { get; set; } = "";
        [JsonProperty("displayName")] public string DisplayName { get; set; } = "";
        [JsonProperty("connectedAt")] public long   ConnectedAt { get; set; }
    }

    // クライアント→サーバー
    public class SigJoin
    {
        [JsonProperty("type")]        public string Type        => "join";
        [JsonProperty("roomId")]      public string RoomId      { get; set; } = "";
        [JsonProperty("role")]        public string Role        { get; set; } = "";
        [JsonProperty("displayName")] public string DisplayName { get; set; } = "";
    }
    public class SigOffer
    {
        [JsonProperty("type")] public string Type => "offer";
        [JsonProperty("to")]   public string To   { get; set; } = "";
        [JsonProperty("sdp")]  public string Sdp  { get; set; } = "";
    }
    public class SigAnswer
    {
        [JsonProperty("type")] public string Type => "answer";
        [JsonProperty("to")]   public string To   { get; set; } = "";
        [JsonProperty("sdp")]  public string Sdp  { get; set; } = "";
    }
    public class SigIce
    {
        [JsonProperty("type")]      public string Type      => "ice";
        [JsonProperty("to")]        public string To        { get; set; } = "";
        [JsonProperty("candidate")] public object Candidate { get; set; } = new();
    }
    public class SigPing
    {
        [JsonProperty("type")] public string Type => "ping";
        [JsonProperty("ts")]   public long   Ts   { get; set; }
    }

    // ─── イベントデータ ───────────────────────────────────────

    public class SigWelcomeEvent
    {
        public string        ClientId { get; set; } = "";
        public string        RoomId   { get; set; } = "";
        public SigPeerInfo[] Peers    { get; set; } = Array.Empty<SigPeerInfo>();
    }
    public class SigOfferEvent  { public string From { get; set; } = ""; public string Sdp { get; set; } = ""; }
    public class SigAnswerEvent { public string From { get; set; } = ""; public string Sdp { get; set; } = ""; }
    public class SigIceEvent    { public string From { get; set; } = ""; public string CandidateJson { get; set; } = ""; }
    public class SigGamePhaseEvent
    {
        public string  Phase        { get; set; } = "";
        public float?  RemainingSec { get; set; }
        public string? Winner       { get; set; }
    }

    // ─── メインクラス ─────────────────────────────────────────

    public class SignalingClient : IDisposable
    {
        public string ClientId { get; private set; } = "";
        public bool   IsConnected => _ws?.IsRunning ?? false;

        // イベント (WSスレッドから発火 → UnityMainThread.Enqueue で安全に処理)
        public event Action<SigWelcomeEvent>?   OnWelcome;
        public event Action<SigPeerInfo>?        OnPeerJoined;
        public event Action<string, string>?     OnPeerLeft;      // (clientId, role)
        public event Action<SigOfferEvent>?      OnOffer;
        public event Action<SigAnswerEvent>?     OnAnswer;
        public event Action<SigIceEvent>?        OnIce;
        public event Action<SigGamePhaseEvent>?  OnGamePhase;
        public event Action<bool>?               OnConnectionChanged;
        public event Action<string>?             OnError;

        private WebsocketClient? _ws;
        private readonly string  _url;
        private readonly string  _roomId;
        private readonly string  _role;
        private readonly string  _displayName;
        private readonly CancellationTokenSource _cts = new();

        // ─── 送信レート制御 (Ping 用) ─────────────────────────
        private float _pingTimer = 0f;

        public SignalingClient(string url, string roomId, string role, string displayName)
        {
            _url         = url;
            _roomId      = roomId;
            _role        = role;
            _displayName = displayName;
        }

        // ─── 接続 ──────────────────────────────────────────────
        public async Task ConnectAsync()
        {
            _ws = new WebsocketClient(new Uri(_url))
            {
                ReconnectTimeout      = TimeSpan.FromSeconds(5),
                ErrorReconnectTimeout = TimeSpan.FromSeconds(5),
                IsReconnectionEnabled = true,
            };

            _ws.MessageReceived.Subscribe(msg =>
            {
                if (msg.Text != null) HandleMessage(msg.Text);
            });

            _ws.ReconnectionHappened.Subscribe(_ =>
            {
                // 再接続時は join を再送
                SendRaw(new SigJoin { RoomId = _roomId, Role = _role, DisplayName = _displayName });
                UnityMainThread.Enqueue(() => OnConnectionChanged?.Invoke(true));
            });

            _ws.DisconnectionHappened.Subscribe(_ =>
            {
                UnityMainThread.Enqueue(() => OnConnectionChanged?.Invoke(false));
            });

            await _ws.Start();
            Debug.Log($"[Signaling] 接続: {_url}");
        }

        // ─── 送信 API ─────────────────────────────────────────
        public void SendOffer(string to, string sdp)   => SendRaw(new SigOffer  { To = to, Sdp = sdp });
        public void SendAnswer(string to, string sdp)  => SendRaw(new SigAnswer { To = to, Sdp = sdp });
        public void SendIce(string to, string candidateJson)
        {
            var candObj = JsonConvert.DeserializeObject(candidateJson) ?? new object();
            SendRaw(new SigIce { To = to, Candidate = candObj });
        }
        public void SendStartGame() => SendRaw(new { type = "start_game" });
        public void SendResetGame() => SendRaw(new { type = "reset_game" });

        // ─── メッセージ解析 ───────────────────────────────────
        private void HandleMessage(string json)
        {
            try
            {
                var obj = JObject.Parse(json);
                var type = obj["type"]?.Value<string>();
                if (type == null) return;

                switch (type)
                {
                    case "welcome":
                        ClientId = obj["clientId"]?.Value<string>() ?? "";
                        var peers = obj["peers"]?.ToObject<SigPeerInfo[]>() ?? Array.Empty<SigPeerInfo>();
                        UnityMainThread.Enqueue(() => OnWelcome?.Invoke(new SigWelcomeEvent
                        {
                            ClientId = ClientId,
                            RoomId   = obj["roomId"]?.Value<string>() ?? "",
                            Peers    = peers,
                        }));
                        break;

                    case "peer_joined":
                        var peer = obj["peer"]?.ToObject<SigPeerInfo>();
                        if (peer != null)
                            UnityMainThread.Enqueue(() => OnPeerJoined?.Invoke(peer));
                        break;

                    case "peer_left":
                        var leftId   = obj["clientId"]?.Value<string>() ?? "";
                        var leftRole = obj["role"]?.Value<string>() ?? "";
                        UnityMainThread.Enqueue(() => OnPeerLeft?.Invoke(leftId, leftRole));
                        break;

                    case "offer":
                        var offerFrom = obj["from"]?.Value<string>() ?? "";
                        var offerSdp  = obj["sdp"]?.Value<string>()  ?? "";
                        UnityMainThread.Enqueue(() => OnOffer?.Invoke(
                            new SigOfferEvent { From = offerFrom, Sdp = offerSdp }));
                        break;

                    case "answer":
                        var ansFrom = obj["from"]?.Value<string>() ?? "";
                        var ansSdp  = obj["sdp"]?.Value<string>()  ?? "";
                        UnityMainThread.Enqueue(() => OnAnswer?.Invoke(
                            new SigAnswerEvent { From = ansFrom, Sdp = ansSdp }));
                        break;

                    case "ice":
                        var iceFrom      = obj["from"]?.Value<string>() ?? "";
                        var iceCandidObj = obj["candidate"];
                        var iceCandJson  = iceCandidObj?.ToString() ?? "{}";
                        UnityMainThread.Enqueue(() => OnIce?.Invoke(
                            new SigIceEvent { From = iceFrom, CandidateJson = iceCandJson }));
                        break;

                    case "game_phase":
                        var gp = new SigGamePhaseEvent
                        {
                            Phase        = obj["phase"]?.Value<string>()        ?? "",
                            RemainingSec = obj["remainingSec"]?.Value<float?>(),
                            Winner       = obj["winner"]?.Value<string?>(),
                        };
                        UnityMainThread.Enqueue(() => OnGamePhase?.Invoke(gp));
                        break;

                    case "pong":
                        // ラテンシ計測（将来実装）
                        break;

                    case "error":
                        var errMsg = $"{obj["code"]?.Value<string>()}: {obj["message"]?.Value<string>()}";
                        Debug.LogWarning($"[Signaling] Error: {errMsg}");
                        UnityMainThread.Enqueue(() => OnError?.Invoke(errMsg));
                        break;
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"[Signaling] HandleMessage: {e.Message}");
            }
        }

        private void SendRaw<T>(T msg)
        {
            if (_ws == null || !_ws.IsRunning) return;
            _ws.Send(JsonConvert.SerializeObject(msg));
        }

        public void Dispose()
        {
            _cts.Cancel();
            _ws?.Dispose();
        }
    }
}

File: projects/plateau-sniper-vs-guardman/packages/web/src/lib/signaling/SignalingClient.ts
// =============================================================
//  lib/signaling/SignalingClient.ts
//  PartyKit WebSocket シグナリングクライアント
//  ICE / SDP のみを扱う。位置データは通さない。
// =============================================================

import type {
  SignalingClientMsg,
  SignalingServerMsg,
  PeerInfo,
  Role,
  GamePhase,
} from "@plateau-sniper/shared"

export type SignalingEventMap = {
  welcome:      { clientId: string; roomId: string; peers: PeerInfo[] }
  peer_joined:  PeerInfo
  peer_left:    { clientId: string; role: Role }
  offer:        { from: string; sdp: string }
  answer:       { from: string; sdp: string }
  ice:          { from: string; candidate: RTCIceCandidateInit }
  game_phase:   { phase: GamePhase; remainingSec?: number; winner?: "sniper" | "bodyguard" }
  connected:    void
  disconnected: void
  error:        { code: string; message: string }
}

type Listener<K extends keyof SignalingEventMap> =
  (data: SignalingEventMap[K]) => void

export class SignalingClient {
  private ws:       WebSocket | null = null
  private handlers: { [K in keyof SignalingEventMap]?: Listener<K>[] } = {}
  private _clientId = ""
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null
  private closed = false

  get clientId() { return this._clientId }

  constructor(
    private readonly url: string,  // wss://...partykit.dev/party/<roomId>
    private readonly role:         Role,
    private readonly displayName:  string,
    private readonly roomId:       string,
  ) {}

  // ─── 接続 ──────────────────────────────────────────────────
  connect() {
    if (this.ws?.readyState === WebSocket.OPEN) return
    this.ws = new WebSocket(this.url)
    this.ws.binaryType = "arraybuffer"

    this.ws.onopen = () => {
      this.emit("connected", undefined as void)
      // 接続直後に join を送る
      this.send({ type: "join", roomId: this.roomId, role: this.role, displayName: this.displayName })
    }

    this.ws.onmessage = (ev) => {
      try {
        const msg: SignalingServerMsg = JSON.parse(ev.data as string)
        this.dispatch(msg)
      } catch {
        console.warn("[Signaling] parse error", ev.data)
      }
    }

    this.ws.onclose = () => {
      this.emit("disconnected", undefined as void)
      if (!this.closed) this.scheduleReconnect()
    }

    this.ws.onerror = (ev) => {
      console.error("[Signaling] ws error", ev)
    }
  }

  disconnect() {
    this.closed = true
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer)
    this.ws?.close()
    this.ws = null
  }

  // ─── シグナリング送信 API ────────────────────────────────────

  sendOffer(to: string, sdp: string)  { this.send({ type: "offer",  to, sdp }) }
  sendAnswer(to: string, sdp: string) { this.send({ type: "answer", to, sdp }) }
  sendIce(to: string, candidate: RTCIceCandidateInit) {
    this.send({ type: "ice", to, candidate })
  }
  sendStartGame()  { this.send({ type: "start_game" }) }
  sendResetGame()  { this.send({ type: "reset_game" }) }

  // ─── イベント登録 ────────────────────────────────────────────
  on<K extends keyof SignalingEventMap>(event: K, fn: Listener<K>) {
    if (!this.handlers[event]) this.handlers[event] = []
    ;(this.handlers[event] as Listener<K>[]).push(fn)
    return this
  }

  off<K extends keyof SignalingEventMap>(event: K, fn: Listener<K>) {
    if (!this.handlers[event]) return
    this.handlers[event] = (this.handlers[event] as Listener<K>[]).filter(h => h !== fn)
  }

  // ─── 内部 ────────────────────────────────────────────────────
  private send(msg: SignalingClientMsg) {
    if (this.ws?.readyState !== WebSocket.OPEN) return
    this.ws.send(JSON.stringify(msg))
  }

  private dispatch(msg: SignalingServerMsg) {
    switch (msg.type) {
      case "welcome":
        this._clientId = msg.clientId
        this.emit("welcome", msg)
        break
      case "peer_joined":   this.emit("peer_joined", msg.peer); break
      case "peer_left":     this.emit("peer_left",   { clientId: msg.clientId, role: msg.role }); break
      case "offer":         this.emit("offer",        { from: msg.from, sdp: msg.sdp }); break
      case "answer":        this.emit("answer",       { from: msg.from, sdp: msg.sdp }); break
      case "ice":           this.emit("ice",          { from: msg.from, candidate: msg.candidate }); break
      case "game_phase":    this.emit("game_phase",   msg); break
      case "error":         this.emit("error",        msg); break
      case "pong":          /* ラテンシ計測は上位で */ break
    }
  }

  private emit<K extends keyof SignalingEventMap>(event: K, data: SignalingEventMap[K]) {
    this.handlers[event]?.forEach(fn => fn(data))
  }

  private scheduleReconnect() {
    this.reconnectTimer = setTimeout(() => {
      console.log("[Signaling] reconnecting...")
      this.connect()
    }, 3000)
  }
}
